game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/fire/general/FiringGroupSplitterGeneralTest.java

```diff
@@ -0,0 +1,357 @@
+package games.strategy.triplea.delegate.battle.steps.fire.general;
+
+import static games.strategy.triplea.Constants.UNIT_ATTACHMENT_NAME;
+import static games.strategy.triplea.delegate.battle.BattleState.Side.DEFENSE;
+import static games.strategy.triplea.delegate.battle.BattleState.Side.OFFENSE;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.AIR_FIRE_NON_SUBS;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.UNITS;
+import static games.strategy.triplea.delegate.battle.FakeBattleState.givenBattleStateBuilder;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenAnyUnit;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitIsAir;
+import static games.strategy.triplea.delegate.battle.steps.BattleStepsTest.givenUnitIsSea;
+import static games.strategy.triplea.delegate.battle.steps.MockGameData.givenGameData;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.battle.steps.fire.FiringGroup;
+import java.util.List;
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class FiringGroupSplitterGeneralTest {
+
+  @Mock GamePlayer attacker;
+  @Mock GamePlayer defender;
+
+  @Test
+  void oneFiringUnitVsOneTargetableUnitMakesOneFiringGroup() {
+    final Unit targetUnit = givenAnyUnit();
+    final Unit fireUnit = givenAnyUnit();
+
+    final List<FiringGroup> firingGroups =
+        FiringGroupSplitterGeneral.of(OFFENSE, unit -> true, UNITS)
+            .apply(
+                givenBattleStateBuilder()
+                    .gameData(givenGameData().withAlliedAirIndependent(true).build())
+                    .attacker(attacker)
+                    .defender(defender)
+                    .attackingUnits(List.of(fireUnit))
+                    .defendingUnits(List.of(targetUnit))
+                    .build());
+
+    assertThat(firingGroups, hasSize(1));
+    assertThat(firingGroups.get(0).getDisplayName(), is(UNITS));
```

Comment:
It seems we are testing multiple behaviors at the same time here. Is the expected display name of units simply because that is the splitter parameter? Have you considered adding a set of tests to only test display name behavior?

Reply:
No, the expected display name will change.  The last two tests in this file test for changed display names: `twoFiringGroupsWithCanNotTarget` and `twoGroupsWhenAirAndSeaVsSub`.

Reply:
I see, what I'm more pointing to is that we are testing multiple/all aspects of firingGroups per test rather than testing on behavior at a time. For example, it's considered a best practice for tests to have only one reason for why they would fail. If for example the firing group size were to change *or* if the displayName were to change, then this test would fail. At a more concrete level, isolating the behaviors makes it easier to see which parameters control which outputs.

Hence I'm suggesting to pivot the tests to group and isolate the tests that verify display name, size, etc.. and do so independently.

Reply:
You'll notice that the majority of the tests don't check for the display name.  Only this one and the other two I mentioned.  I understand that the test should generally only have one reason why they should fail.  And I believe this test is structured that way.

Reply:
Can you clarify "test should generally only have one reason why they should fail. And I believe this test is structured that way."

From an initial look, it seems like if the firing group size changes this test would fail, or if the display name changes it would fail. The size of the firing group and display name are independent of one another, which makes me think that if one of two independent things were to either change, then this test method would fail for either reason. Are the two values, the firing group size and the display name actually linked?

Reply:
This test is validating what a one firing group result should look like.  And a single firing group result will always have the same name.  This is a display name and the only reason to change the display name that is passed in is if there are multiple groups that need to be differentiated.